### PR TITLE
Leaves - Mariya, Cloudy

### PR DESCRIPTION
**Pull request by M-Burr**
# Video Store API
Congratulations! You're submitting your assignment!
If you didn't get to the functionality the question is asking about, reply with what you would have done if you had completed it.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Explain how you came up with the design of your ERD, based on the seed data. | After reading the requirements and creating our ERD, we hoped to have rentals be a joined table without a model, but after working through wave 1, we decided rentals should be its own model so that we could add business logic to rentals. In summary, our original ERD was wrong and we adapted as we progressed through the project.
What would be the Big-O time complexity of your /customers & /movies endpoints? What does the time complexity depend on? Explain your reasoning. | The O time complexity for customers and movies is: O(n), where "n" represents the total number of customer objects or movie objects (i.e., the hashes in the array in json). The amount of time the function takes is proportional to n items (i.e., it is a linear relationship).
What is the Big-O time complexity of the `POST /rentals/check-in` endpoint? What does the time complexity depend on? Explain your reasoning.  | O(1) bc it always display an error (to us, not postman). This result is constant, regardless of the size of the input, therefore it is O(1) (constant).
Describe a set of positive and negative test cases you implemented for a model. | We tested model validations to see if requirements were present.
Describe a set of positive and negative test cases you implemented for a controller. | We tested to see if a rental checkout was successful and confirm that a rental would fail with invalid data.
How does your API respond when bad data is sent to it? | returns a hash where the key("ok") has a value of false and they key("errors") has a description of what errors are present
Describe one of your custom model methods and why you chose to wrap that functionality into a method. | We did not use custom model methods.
Do you have any recommendations on how we could improve this project for the next cohort? | n/a - great project!
Link to Trello | NA
Link to ERD | We did it on paper
Summary | It was a fun project and we learned a lot. Thanks!

### CODE REVIEW

File: app/controllers/rentals_controller.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      render json: rental.as_json(only: [:customer_id, :movie_id]), status: :ok
+      return
+    else
+      # binding.pry
+      render json: { ok: false, "errors" => ["Not Found"]}, status: :not_found
+      return
+    end
+  end
+
+  def checkin
```
Comment (beccaelenzil) [inline app/controllers/rentals_controller.rb:16]: Consider implementing Checkin so that it does not create new rental but checks in an existing rental. You'd need to first find the rental and then checkin. Some of this logic would go in a custom model method.

Comment (beccaelenzil) [reply app/controllers/rentals_controller.rb:16]: In addition, checkin and checkout should affect available movie inventory -- this will likely be a custom method in the movie model.

File: app/controllers/movies_controller.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  end
+  
+  def create
+    movie = Movie.new(title: params[:title], overview: params[:overview], release_date:params[:release_date], inventory:params[:inventory])
+    
+    if movie.save 
+      render json: movie.as_json(only: [:id, :title, :overview, :release_date, :inventory]), status: :ok
+      return 
+    else 
+      render json: { ok: false, "errors" => ["Not Found"]}, status: :not_found
```
Comment (beccaelenzil) [inline app/controllers/movies_controller.rb:28]: Instead of `"error" --> ["Not Found"]` you should report any errors in the movie model created by validations by accessing `movie.errors.messages`.

File: test/models/customer_test.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+describe Customer do
+  describe "relationships" do 
+    it "has movies" do 
+      # Arrange
+      mariya = customers(:mariya)
+      up = movies(:up)
+      avengers = movies(:avengers)
+      zootopia = movies(:zootopia)
+      
+      #Act
```
Comment (beccaelenzil) [inline test/models/customer_test.rb:12]: It seems you don't use the variable `rental`. As such, you don't need this assignment.

File: test/models/customer_test.rb [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      # Arrange
+      mariya = customers(:mariya)
+      up = movies(:up)
+      
+      # Act
+      rental = Rental.create(customer_id: mariya.id, movie_id: up.id, check_out_date: Date.today, due_date: Date.today + 7, check_in_date: nil)
+      
+      # Assert
+      expect(mariya.movies.count).must_equal 1
+    end
```
Comment (beccaelenzil) [inline test/models/customer_test.rb:32]: Consider the case where a movie has no rentals or customers.

File: test/models/customer_test.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  end
+  
+  describe "validations" do
```
Comment (beccaelenzil) [inline test/models/customer_test.rb:35]: thorough validation tests. nice work!

File: test/models/movie_test.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      rental4 = Rental.create(customer_id: sara.id, movie_id: @up.id, check_out_date: Date.today, due_date: Date.today + 7, check_in_date: nil)
+    end
+    
+    it "has customers" do 
+      expect(@up.customers.count).must_equal 4
+    end
+    
+    it "has rentals" do
+      expect(@up.rentals.count).must_equal 4
+    end
```
Comment (beccaelenzil) [inline test/models/movie_test.rb:23]: Also consider the edge cases of having no rentals or customers.

**Comment by beccaelenzil** [comment]
# Video Store
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene |yes
Comprehension questions |yes
**General** |
Business Logic in Models |business logic to update available inventory is missing (from model or controller)
All required endpoints return expected JSON data |yes
Requests respond with appropriate HTTP Status Code |yes
Errors are reported |see comment
**Testing** |
Passes all Smoke Tests |wave 2- all passed, wave 3- 3 failures (1) Customer's movie check out count increased | AssertionError: expected false to be truthy, (2) Movie available_inventory decreased | AssertionError: expected false to be truthy, (3) Check In Movie: Status code is 200 | AssertionError: expected false to be truthy
Model Tests - all relations, validations, and custom functions test positive & negative cases | consider a few additional edge cases for relationships
Controller Tests - URI parameters and data in the request body have positive & negative cases | yes
**Overall** | Good work creating a web API with model relationships and validations. I am impressed by the thoroughness of your controller tests. You are missing the key feature of updating the available inventory with the checkin and checkout actions. Take a look at my inline comments, and the requirements and smoke tests with regards to this feature, and let me know what questions you have.

File: app/controllers/rentals_controller.rb (hunk 1 above)
Comment (beccaelenzil) [reply app/controllers/rentals_controller.rb:16]: I realize now that movie inventory is optional -- I talked about this with Cloudy
